Update to align with stage 3 conditions

rbuckton opened the pull request:
During the January TC39 meeting, the following changes are required to reach consensus for Stage 3:

- `new.target` should follow the same behavior as methods and field initializers and return `undefined` (#25).
- A class may have multiple `static {}` blocks which are evaluated in document order interleaved with static field initializers (#26).

In addition, I've updated the specification text to be a delta from the [Static Class Features proposal](https://arai-a.github.io/ecma262-compare/?pr=1668).

In addition the following are required:
- Additional review by the ECMA-262 editors
- Additional review by @littledan

rbuckton committed:
Update to align with stage 3 conditions

rbuckton commented:
@littledan: You requested that I investigate leveraging `Block` rather than a custom parse node for `ClassStaticBlock` (i.e., `` `static` Block ``). I'm not sure this is feasible, as function bodies require additional semantics that aren't supported by the Static and Runtime semantics of _Block_, otherwise we would also just be reusing _Block_ for productions like [_FunctionDeclaration_](https://tc39.es/ecma262/#prod-FunctionDeclaration), et. al. 

I have, however, leveraged FunctionCreate and MakeMethod, and removed the custom class static block environment record in an effort to simplify the specification text.

michaelficarra commented on spec/index.html:
```diff
@@ -4,18 +4,25 @@
 <pre class="metadata">
 title: ECMAScript class static initialization blocks
-stage: 1
+stage: 2
```
```suggestion
stage: 3
```

rbuckton committed:
Update spec/index.html

Co-authored-by: Michael Ficarra <github@michael.ficarra.me>

syg commented on spec/sec-class-definitions-patch.html:
```diff
@@ -5,6 +5,8 @@ <h2>Syntax</h2>
       MethodDefinition[?Yield, ?Await]
       `static` MethodDefinition[?Yield, ?Await]
+      FieldDefinition[?Yield, ?Await] `;`
+      staticFieldDefinition[?Yield, ?Await] `;`
```
```suggestion
      `static` FieldDefinition[?Yield, ?Await] `;`
```

syg commented on spec/sec-class-definitions-patch.html:
```diff
@@ -37,19 +33,25 @@ <h1>Static Semantics: Early Errors</h1>
+        <p><ins>It is a Syntax Error if any element of the LexicallyDeclaredNames of |ClassStaticBlockStatementList| also occurs in the VarDeclaredNames of |ClassStaticBlockStatementList|.</ins></p>
+      </li>
+      <li>
+        <p><ins>It is a Syntax Error ContainsDuplicateLabels of |ClassStaticBlockStatementList| with argument &laquo; &raquo; is *true*.</ins></p>
```
```suggestion
        <p><ins>It is a Syntax Error if ContainsDuplicateLabels of |ClassStaticBlockStatementList| with argument &laquo; &raquo; is *true*.</ins></p>
```

syg commented on spec/sec-class-definitions-patch.html:
```diff
@@ -37,19 +33,25 @@ <h1>Static Semantics: Early Errors</h1>
       </li>
       <li>
-        <p><ins>It is a Syntax Error if any element of the LexicallyDeclaredNames of ClassStaticBlockStatementList also occurs in the VarDeclaredNames of ClassStaticBlockStatementList.</ins></p>
+        <p><ins>It is a Syntax Error ContainsUndefinedBreakTarget of |ClassStaticBlockStatementList| with argument &laquo; &raquo; is *true*.</ins></p>
```
```suggestion
        <p><ins>It is a Syntax Error if ContainsUndefinedBreakTarget of |ClassStaticBlockStatementList| with argument &laquo; &raquo; is *true*.</ins></p>
```

syg commented on spec/sec-class-definitions-patch.html:
```diff
@@ -37,19 +33,25 @@ <h1>Static Semantics: Early Errors</h1>
       </li>
       <li>
-        <p><ins>It is a Syntax Error ContainsDuplicateLabels of ClassStaticBlockStatementList with argument &laquo; &raquo; is *true*.</ins></p>
+        <p><ins>It is a Syntax Error ContainsUndefinedContinueTarget of |ClassStaticBlockStatementList| with arguments &laquo; &raquo; and &laquo; &raquo; is *true*.</ins></p>
```
```suggestion
        <p><ins>It is a Syntax Error if ContainsUndefinedContinueTarget of |ClassStaticBlockStatementList| with arguments &laquo; &raquo; and &laquo; &raquo; is *true*.</ins></p>
```

syg commented on spec/sec-class-definitions-patch.html:
```diff
@@ -37,19 +33,25 @@ <h1>Static Semantics: Early Errors</h1>
       </li>
       <li>
-        <p><ins>It is a Syntax Error ContainsUndefinedBreakTarget of ClassStaticBlockStatementList with argument &laquo; &raquo; is *true*.</ins></p>
+        <p><ins>It is a Syntax Error ContainsArguments of |ClassStaticBlockStatementList| is *true*.</ins></p>
```
```suggestion
        <p><ins>It is a Syntax Error if ContainsArguments of |ClassStaticBlockStatementList| is *true*.</ins></p>
```

syg commented on spec/sec-class-definitions-patch.html:
```diff
@@ -37,19 +33,25 @@ <h1>Static Semantics: Early Errors</h1>
       </li>
       <li>
-        <p><ins>It is a Syntax Error ContainsUndefinedContinueTarget of ClassStaticBlockStatementList with arguments &laquo; &raquo; and &laquo; &raquo; is *true*.</ins></p>
+        <p><ins>It is a Syntax Error |ClassStaticBlockStatementList| Contains |SuperCall| is *true*.</ins></p>
```
```suggestion
        <p><ins>It is a Syntax Error if |ClassStaticBlockStatementList| Contains |SuperCall| is *true*.</ins></p>
```

syg commented on spec/sec-class-definitions-patch.html:
```diff
@@ -171,9 +156,31 @@ <h1>Static Semantics: VarScopedDeclarations</h1>
+
+  <emu-clause id="sec-runtime-semantics-classelementevaluation">
+    <h1>Runtime Semantics: ClassElementEvaluation</h1>
+    <p>With parameters _object_ and _enumerable_.</p>
```
What's the _enumerable_ for?

rbuckton replied:
ClassElementEvaluation takes two parameters in the class fields proposal. Note that this is existing spec text (no `<ins>`), though I've elided the existing SDO algorithms for brevity to focus on what this proposal changes (the `<ins>` blocks on the following lines).

syg commented on spec/sec-class-definitions-patch.html:
```diff
@@ -171,9 +156,31 @@ <h1>Static Semantics: VarScopedDeclarations</h1>
   </emu-clause>
 
+  <emu-clause id="sec-runtime-semantics-classstaticblockevaluation">
+    <h1><ins>Runtime Semantics: ClassStaticBlockEvaluation</ins></h1>
```
This would be clearer if it were named ClassStaticBlockDefinitionEvaluation.

syg commented on spec/sec-ecmascript-specification-types-patch.html:
```diff
@@ -0,0 +1,36 @@
+        </tr>
+        <tr>
+          <td>
+            [[Block]]
```
I am unreasonably shook by a field named [[Block]] that contains a function object. Any ideas for a less misleading name?

rbuckton replied:
`[[StaticBlock]]`, or maybe `[[Body]]`?

syg reviewed (approved):
lgtm, I think all my comments thus far are editorial.

ljharb reviewed (approved):
LGTM pending shu's comments

rbuckton committed:
Apply suggestions from code review

Co-authored-by: Shu-yu Guo <shu@rfrn.org>

rbuckton committed:
PR Feedback

rbuckton commented:
@syg: I've made updates based on your suggestions. Please let me know if there's anything else you see that I should address.

@littledan: When you have some time, please provide your feedback.

littledan commented on spec/biblio.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "https://tc39.es/ecma262/": [
```
Maybe this should link to the diff for now, so it is not a dead link?

rbuckton replied:
Changing it to `https://arai-a.github.io/ecma262-compare/?pr=1668` didn't fix the link. As soon as I open the diff it ignores the `#sec-class-definitions-static-semantics-containsarguments` added by ecmarkup.

Using `https://arai-a.github.io/ecma262-compare/history/PR/1668/94b1af23d6ca075c6635bcd5b96c6450ec580f51` seems to work, though, but will get out of date if new commits are added.

rbuckton replied:
I can kind of hack around it by using this biblio entry:
```
{
    "https://arai-a.github.io/ecma262-compare/?pr=1668&id=sec-class-definitions-static-semantics-containsarguments": [
        {"type": "clause", "id": "sec-class-definitions-static-semantics-containsarguments", ... }
    ]
}
```
Since that diff tool uses `id=`.

littledan reviewed (approved):
This PR is perfect--it encodes the semantics we agreed on, and removes the duplication that I was most concerned about. Thanks for the update, and apologies for my delay.

rbuckton committed:
PR feedback, updated ecma262biblio for algorithm links

rbuckton commented:
I updated the biblio entry above, and also generated a fresh `ecma262biblio.json` (would be nice if this [were automatic](https://github.com/tc39/ecma262/issues/2321)) and fixed some algorithm references where the names have changed (`FunctionCreate`->`OrdinaryFunctionCreate`).

syg reviewed (approved):
lgtm!
